### ISSUE TITLE
v2: chain middleware removes headers added by headers middleware

### ISSUE THREAD

**Issue by JamesKyburz** [containous/traefik#5890]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

Expect to see custom headers when the headers middleware is used.

### What did you expect to see?

custom headers

### What did you see instead?

When the chain middleware is used custom headers are not present in the response.

### Output of `traefik version`: (_What version of Traefik are you using?_)

v2.1

```
(paste your output here)
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```toml
# (paste your configuration here)
```

Using chain the custom headers are not present.

Example docker-compose-bug.yml

~/Documents/src/traefik-bug/root1 ᐅ cat docker-compose-bug.yml
```yml
version: '3'
services:
  reverse-proxy:
    image: traefik:v2.1
    networks:
      - traefik-bug-root1
    command:
      - '--providers.docker'
    ports:
    - 8095:80
    volumes:
    - /var/run/docker.sock:/var/run/docker.sock
    restart: always

  test:
    networks:
    - traefik-bug-root1
    image: hashicorp/http-echo
    command: "-listen :5000 -text='hi'"
    labels:
    - 'traefik.enable=true'
    - 'traefik.http.routers.test.rule=PathPrefix(`/test`)'
    - 'traefik.http.routers.test.middlewares=test'
    - 'traefik.http.services.test.loadbalancer.server.port=5000'
    - 'traefik.http.middlewares.test.chain.middlewares=test-headers,test-compress'
    - 'traefik.http.middlewares.test-compress.compress=true'
    - 'traefik.http.middlewares.test-headers.headers.customresponseheaders.X-Custom-Response-Header=value'

networks:
  traefik-bug-root1:
    driver: bridge
```

If the chain middleware is not used custom headers are present.

~/Documents/src/traefik-bug/root1 ᐅ cat docker-compose-ok.yml
```yml
version: '3'
services:
  reverse-proxy:
    image: traefik:v2.1
    networks:
      - traefik-bug-root1
    command:
      - '--providers.docker'
    ports:
    - 8095:80
    volumes:
    - /var/run/docker.sock:/var/run/docker.sock
    restart: always

  test:
    networks:
    - traefik-bug-root1
    image: hashicorp/http-echo
    command: "-listen :5000 -text='hi'"
    labels:
    - 'traefik.enable=true'
    - 'traefik.http.routers.test.rule=PathPrefix(`/test`)'
    - 'traefik.http.routers.test.middlewares=test-headers'
    - 'traefik.http.services.test.loadbalancer.server.port=5000'
    - 'traefik.http.middlewares.test-headers.headers.customresponseheaders.X-Custom-Response-Header=value'

networks:
  traefik-bug-root1:
    driver: bridge
```

### If applicable, please paste the log output in DEBUG level (`--log.level=DEBUG` switch)

```
(paste your output here)
```

**Comment by juliens** [comment]
Could you provide us the logs in debug mode, and could you enable the `--api.insecure` and send us the result of `http://127.0.0.1:8080/api/rawdata`

**Comment by robinwangrubin** [comment]
Yes, I met the same problem. When i used the middleware chain.  The `customresponseheaders` doesn't work.

**Comment by dtomcej** [comment]
I think the issue may come from the `responseModifiers` not being properly applied when the headers middleware is included in a chain:

**Comment by Dmry** [comment]
Ran into the same issue. In case it's still helpful, please find my raw data from http://127.0.0.1:8080/api/rawdata attached.

[rawdata.txt](https://github.com/containous/traefik/files/4022495/rawdata.txt)

**Comment by pascalandy** [comment]
Hello folks, 
`chain` middleware is working as expected for me.

### compose nginx

```
    labels:
	...
      - "traefik.http.routers.nginx.middlewares=RuleGrpMain"
```

### compose traefik

```
-
    labels:
	...
"traefik.http.middlewares.RuleGrpMain.chain.middlewares=compress_all,ratelimit_all,retry_all,header_all,slash_step1,slash_step2"
```

### logs

![5890_15](https://user-images.githubusercontent.com/6694151/72562941-9aa6a700-387a-11ea-92af-35034e737906.jpg)
![5890_11](https://user-images.githubusercontent.com/6694151/72562943-9aa6a700-387a-11ea-9c79-7b358a43e241.jpg)
![5890_12](https://user-images.githubusercontent.com/6694151/72562944-9aa6a700-387a-11ea-8578-00b5357ab935.jpg)
![5890_13](https://user-images.githubusercontent.com/6694151/72562945-9b3f3d80-387a-11ea-8e11-353bcc46e59f.jpg)
![5890_14](https://user-images.githubusercontent.com/6694151/72562946-9b3f3d80-387a-11ea-920d-a6a610b17b4d.jpg)

**Comment by pascalandy** [comment]
Here is a funny behavior. I have 3 webapps (whoami, nginx, caddy). All of them are using the same chain:

```
  #### Apply rules (middlewares)
  - "traefik.http.routers.home.middlewares=RuleGrpMain"
```

I cannot explain why caddy has no headers at all 
(You can see my domain is also behind Cloudflare.)

### whoami

```
➜  / curl -I https://devkiwi.club
HTTP/2 200
date: Fri, 17 Jan 2020 16:09:10 GMT
content-type: text/plain; charset=utf-8
content-length: 483
set-cookie: __cfduid=d96da8169317396e0c45bd3719987421f1579277350; expires=Sun, 16-Feb-20 16:09:10 GMT; path=/; domain=.devkiwi.club; HttpOnly; SameSite=Lax
content-security-policy: default-src https:
feature-policy: accelerometer 'none'; ambient-light-sensor 'none'; camera 'none'; geolocation 'none'; gyroscope 'none'; magnetometer 'none'; microphone 'none'; usb 'none'; midi 'none'; sync-xhr 'none'; vr 'none'
referrer-policy: strict-origin-when-cross-origin
strict-transport-security: max-age=31536000; includeSubDomains; preload
vary: Accept-Encoding
x-content-type-options: nosniff
x-frame-options: DENY
x-xss-protection: 1; mode=block
cf-cache-status: DYNAMIC
expect-ct: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
server: cloudflare
cf-ray: 5569978f3a7d3b4e-YVR
```

### nginx

```
➜  / curl -I https://devkiwi.club/nginx/
HTTP/2 200
date: Fri, 17 Jan 2020 16:06:48 GMT
content-type: text/html
set-cookie: __cfduid=d647e69770eefc87cdafcfdb4f5b53f511579277208; expires=Sun, 16-Feb-20 16:06:48 GMT; path=/; domain=.devkiwi.club; HttpOnly; SameSite=Lax
accept-ranges: bytes
content-security-policy: default-src https:
feature-policy: accelerometer 'none'; ambient-light-sensor 'none'; camera 'none'; geolocation 'none'; gyroscope 'none'; magnetometer 'none'; microphone 'none'; usb 'none'; midi 'none'; sync-xhr 'none'; vr 'none'
last-modified: Mon, 13 Jan 2020 14:50:45 GMT
referrer-policy: strict-origin-when-cross-origin
strict-transport-security: max-age=31536000; includeSubDomains; preload
vary: Accept-Encoding
x-content-type-options: nosniff
x-frame-options: DENY
x-xss-protection: 1; mode=block
cf-cache-status: DYNAMIC
expect-ct: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
server: cloudflare
cf-ray: 556994175a1ff4b2-YVR
```

### caddy

```
➜  / curl -I https://devkiwi.club/caddy/
HTTP/2 200
date: Fri, 17 Jan 2020 16:08:36 GMT
content-type: text/html; charset=utf-8
set-cookie: __cfduid=d40191a814363abcda73c7c498d55b0de1579277316; expires=Sun, 16-Feb-20 16:08:36 GMT; path=/; domain=.devkiwi.club; HttpOnly; SameSite=Lax
accept-ranges: bytes
last-modified: Mon, 13 Jan 2020 14:50:45 GMT
vary: Accept-Encoding
cf-cache-status: DYNAMIC
expect-ct: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
server: cloudflare
cf-ray: 556996baabb0f4be-YVR
```

**Comment by pascalandy** [comment]
Well, I think the official caddy image is overriding the headers. Could it be possible?

Using a custom image of caddy (devmtl/caddyfire:1.0.3), where I defined the headers in Caddy's config) and it's all good.

**Comment by JamesKyburz** [comment]
```sh
cat docker-compose-bug.yml
```

```yaml
version: '3'
services:
  reverse-proxy:
    image: traefik:v2.1
    networks:
      - traefik-bug-root1
    command:
      - '--providers.docker'
      - '--api.insecure'
    ports:
    - 8095:80
    - 8080:8080
    volumes:
    - /var/run/docker.sock:/var/run/docker.sock
    restart: always

  test:
    networks:
    - traefik-bug-root1
    image: hashicorp/http-echo
    command: "-listen :5000 -text='hi'"
    labels:
    - 'traefik.enable=true'
    - 'traefik.http.routers.test.rule=PathPrefix(`/test`)'
    - 'traefik.http.routers.test.middlewares=test'
    - 'traefik.http.services.test.loadbalancer.server.port=5000'
    - 'traefik.http.middlewares.test.chain.middlewares=test-headers,test-compress'
    - 'traefik.http.middlewares.test-compress.compress=true'
    - 'traefik.http.middlewares.test-headers.headers.customresponseheaders.X-Custom-Response-Header=value'

networks:
  traefik-bug-root1:
    driver: bridge

```

```sh
docker-compose -f docker-compose-bug.yml up 
```
```sh
http --headers http://localhost:8095/test              
HTTP/1.1 200 OK
Content-Length: 3
Content-Type: text/plain; charset=utf-8
Date: Sat, 18 Jan 2020 14:26:08 GMT
Vary: Accept-Encoding
X-App-Name: http-echo
X-App-Version: 0.2.3
```

```sh
curl http://localhost:8080/api/rawdata | jq ''
```

```json
{
  "routers": {
    "api@internal": {
      "entryPoints": [
        "traefik"
      ],
      "service": "api@internal",
      "rule": "PathPrefix(`/api`)",
      "priority": 2147483646,
      "status": "enabled",
      "using": [
        "traefik"
      ]
    },
    "dashboard@internal": {
      "entryPoints": [
        "traefik"
      ],
      "middlewares": [
        "dashboard_redirect@internal",
        "dashboard_stripprefix@internal"
      ],
      "service": "dashboard@internal",
      "rule": "PathPrefix(`/`)",
      "priority": 2147483645,
      "status": "enabled",
      "using": [
        "traefik"
      ]
    },
    "reverse-proxy-root1@docker": {
      "service": "reverse-proxy-root1",
      "rule": "Host(`reverse-proxy-root1`)",
      "status": "enabled",
      "using": [
        "http",
        "traefik"
      ]
    },
    "test@docker": {
      "middlewares": [
        "test@docker"
      ],
      "service": "test",
      "rule": "PathPrefix(`/test`)",
      "status": "enabled",
      "using": [
        "http",
        "traefik"
      ]
    }
  },
  "middlewares": {
    "dashboard_redirect@internal": {
      "redirectRegex": {
        "regex": "^(http:\\/\\/[^:\\/]+(:\\d+)?)\\/$",
        "replacement": "${1}/dashboard/",
        "permanent": true
      },
      "status": "enabled",
      "usedBy": [
        "dashboard@internal"
      ]
    },
    "dashboard_stripprefix@internal": {
      "stripPrefix": {
        "prefixes": [
          "/dashboard/",
          "/dashboard"
        ]
      },
      "status": "enabled",
      "usedBy": [
        "dashboard@internal"
      ]
    },
    "test-compress@docker": {
      "compress": {},
      "status": "enabled"
    },
    "test-headers@docker": {
      "headers": {
        "customResponseHeaders": {
          "X-Custom-Response-Header": "value"
        }
      },
      "status": "enabled"
    },
    "test@docker": {
      "chain": {
        "middlewares": [
          "test-headers@docker",
          "test-compress@docker"
        ]
      },
      "status": "enabled",
      "usedBy": [
        "test@docker"
      ]
    }
  },
  "services": {
    "api@internal": {
      "status": "enabled",
      "usedBy": [
        "api@internal"
      ]
    },
    "dashboard@internal": {
      "status": "enabled",
      "usedBy": [
        "dashboard@internal"
      ]
    },
    "reverse-proxy-root1@docker": {
      "loadBalancer": {
        "servers": [
          {
            "url": "http://172.27.0.3:80"
          }
        ],
        "passHostHeader": true
      },
      "status": "enabled",
      "usedBy": [
        "reverse-proxy-root1@docker"
      ],
      "serverStatus": {
        "http://172.27.0.3:80": "UP"
      }
    },
    "test@docker": {
      "loadBalancer": {
        "servers": [
          {
            "url": "http://172.27.0.2:5000"
          }
        ],
        "passHostHeader": true
      },
      "status": "enabled",
      "usedBy": [
        "test@docker"
      ],
      "serverStatus": {
        "http://172.27.0.2:5000": "UP"
      }
    }
  }
}


```

**Comment by JamesKyburz** [comment]
```sh
cat docker-compose.ok.yaml
```

```yaml
version: '3'
services:
  reverse-proxy:
    image: traefik:v2.1
    networks:
      - traefik-bug-root1
    command:
      - '--providers.docker'
      - '--api.insecure'
    ports:
    - 8095:80
    - 8080:8080
    volumes:
    - /var/run/docker.sock:/var/run/docker.sock
    restart: always

  test:
    networks:
    - traefik-bug-root1
    image: hashicorp/http-echo
    command: "-listen :5000 -text='hi'"
    labels:
    - 'traefik.enable=true'
    - 'traefik.http.routers.test.rule=PathPrefix(`/test`)'
    - 'traefik.http.routers.test.middlewares=test-headers'
    - 'traefik.http.services.test.loadbalancer.server.port=5000'
    - 'traefik.http.middlewares.test-headers.headers.customresponseheaders.X-Custom-Response-Header=value'

networks:
  traefik-bug-root1:
    driver: bridge
```
```sh
docker-compose -f docker-compose-ok.yml up
```

```sh
http --headers http://localhost:8095/test              
HTTP/1.1 200 OK
Content-Length: 3
Content-Type: text/plain; charset=utf-8
Date: Sat, 18 Jan 2020 14:28:55 GMT
X-App-Name: http-echo
X-App-Version: 0.2.3
X-Custom-Response-Header: value
```

```sh
curl http://localhost:8080/api/rawdata | jq ''
```

```json
{
  "routers": {
    "api@internal": {
      "entryPoints": [
        "traefik"
      ],
      "service": "api@internal",
      "rule": "PathPrefix(`/api`)",
      "priority": 2147483646,
      "status": "enabled",
      "using": [
        "traefik"
      ]
    },
    "dashboard@internal": {
      "entryPoints": [
        "traefik"
      ],
      "middlewares": [
        "dashboard_redirect@internal",
        "dashboard_stripprefix@internal"
      ],
      "service": "dashboard@internal",
      "rule": "PathPrefix(`/`)",
      "priority": 2147483645,
      "status": "enabled",
      "using": [
        "traefik"
      ]
    },
    "reverse-proxy-root1@docker": {
      "service": "reverse-proxy-root1",
      "rule": "Host(`reverse-proxy-root1`)",
      "status": "enabled",
      "using": [
        "http",
        "traefik"
      ]
    },
    "test@docker": {
      "middlewares": [
        "test-headers@docker"
      ],
      "service": "test",
      "rule": "PathPrefix(`/test`)",
      "status": "enabled",
      "using": [
        "http",
        "traefik"
      ]
    }
  },
  "middlewares": {
    "dashboard_redirect@internal": {
      "redirectRegex": {
        "regex": "^(http:\\/\\/[^:\\/]+(:\\d+)?)\\/$",
        "replacement": "${1}/dashboard/",
        "permanent": true
      },
      "status": "enabled",
      "usedBy": [
        "dashboard@internal"
      ]
    },
    "dashboard_stripprefix@internal": {
      "stripPrefix": {
        "prefixes": [
          "/dashboard/",
          "/dashboard"
        ]
      },
      "status": "enabled",
      "usedBy": [
        "dashboard@internal"
      ]
    },
    "test-headers@docker": {
      "headers": {
        "customResponseHeaders": {
          "X-Custom-Response-Header": "value"
        }
      },
      "status": "enabled",
      "usedBy": [
        "test@docker"
      ]
    }
  },
  "services": {
    "api@internal": {
      "status": "enabled",
      "usedBy": [
        "api@internal"
      ]
    },
    "dashboard@internal": {
      "status": "enabled",
      "usedBy": [
        "dashboard@internal"
      ]
    },
    "reverse-proxy-root1@docker": {
      "loadBalancer": {
        "servers": [
          {
            "url": "http://172.27.0.2:80"
          }
        ],
        "passHostHeader": true
      },
      "status": "enabled",
      "usedBy": [
        "reverse-proxy-root1@docker"
      ],
      "serverStatus": {
        "http://172.27.0.2:80": "UP"
      }
    },
    "test@docker": {
      "loadBalancer": {
        "servers": [
          {
            "url": "http://172.27.0.3:5000"
          }
        ],
        "passHostHeader": true
      },
      "status": "enabled",
      "usedBy": [
        "test@docker"
      ],
      "serverStatus": {
        "http://172.27.0.3:5000": "UP"
      }
    }
  }
}
```

**Comment by JamesKyburz** [comment]
@pascalandy I don't think hashicorp/http-echo modifies the headers, I have a working example with `docker-compose-ok.yml` which doesn't use the chain middleware, the problem is only with the chain middleware as shown in `docker-compose-bug.yml`

**Comment by pascalandy** [comment]
There is a [similar use case on the forum](https://community.containo.us/t/cors-headers-are-broken-for-me-since-2-0-6/3756). It magically works when you put your headers in the beginning of the chain.

**Comment by JamesKyburz** [comment]
@pascalandy That didn't work with my `docker-compose-bug.yml` example.

Neither 
`'traefik.http.middlewares.test.chain.middlewares=test-headers,test-compress'`
nor
`'traefik.http.middlewares.test.chain.middlewares=test-compress,test-headers'`

worked.

**Comment by pascalandy** [comment]
> There is a similar use case on the forum. It magically works when you put your headers in the beginning of the chain.

- It fixed the headers with my Caddy's container. But why?
- Still no headers with http-echo's container.

**Comment by JamesKyburz** [comment]
I also tried `'traefik.http.middlewares.test.chain.middlewares=test-headers'` and that didn't work either.

So the problem seems to be that the `chain` middleware doesn't respect custom headers, and has nothing to do with order of multiple middlewares.

**Comment by olivierHa** [comment]
Same issue here. 
If used outside a chain (directly attached to a router), headers middleware is working as expected.
But within a chain, headers middleware doesn't work at all.

**Comment by traefiker** [comment]
Closed by #6233.
